File: assignments/assignment-1-solution/src/CrazyEights/CardDeck/DiscardPile.cs
using CrazyEights.Cards;
using CrazyEights.Domain;

namespace CrazyEights.CardDeck;

/// <summary>
/// Represents the discard pile and active suit state.
/// </summary>
public class DiscardPile
{
    private readonly Stack<ICard> _discardedCards = [];

    public ICard TopCard
    {
        get
        {
            if (_discardedCards.Count == 0)
                throw new InvalidOperationException("Discard pile is empty.");

            return _discardedCards.Peek();
        }
    }

    public SuitType ActiveSuit => OverriddenSuit == SuitType.NotSet ? TopCard.Suit : OverriddenSuit;

    private SuitType OverriddenSuit { get; set; } = SuitType.NotSet;

    public void AddCard(ICard card)
    {
        _discardedCards.Push(card);
        SetTopCardAsActiveSuit();
    }

    private void SetTopCardAsActiveSuit()
    {
        OverriddenSuit = SuitType.NotSet;
    }

    public void OverrideTopCardSuit(SuitType newSuit)
    {
        if (TopCard.Suit == newSuit)
        {
            SetTopCardAsActiveSuit();
            return;
        }

        OverriddenSuit = newSuit;
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/carddeck/DiscardPile.java
package edu.kennesaw.crazy8s.carddeck;

import edu.kennesaw.crazy8s.cards.Card;
import edu.kennesaw.crazy8s.domain.SuitType;
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Represents the discard pile and active suit state.
 */
public class DiscardPile {
    private final Deque<Card> discardedCards = new ArrayDeque<>();
    private SuitType overriddenSuit = SuitType.NOT_SET;

    public Card getTopCard() {
        if (discardedCards.isEmpty()) {
            throw new IllegalStateException("Discard pile is empty.");
        }

        return discardedCards.peek();
    }

    public SuitType getActiveSuit() {
        return overriddenSuit == SuitType.NOT_SET ? getTopCard().getSuit() : overriddenSuit;
    }

    public void addCard(Card card) {
        discardedCards.push(card);
        setTopCardAsActiveSuit();
    }

    private void setTopCardAsActiveSuit() {
        overriddenSuit = SuitType.NOT_SET;
    }

    public void overrideTopCardSuit(SuitType newSuit) {
        if (getTopCard().getSuit() == newSuit) {
            setTopCardAsActiveSuit();
            return;
        }

        overriddenSuit = newSuit;
    }
}
